### PR TITLE
odo login

### PR DESCRIPTION
**Pull request by surajnarwade**
add command `odo login`

as mentioned in https://github.com/redhat-developer/odo/issues/555

### CODE REVIEW

**Comment by surajnarwade** [comment]
can you please review or tag respective person:

cc @kadel @jorgemoralespou  @simo5 @enj

**Comment by surajnarwade** [comment]
travis is failing because of license check

**Comment by veillard** [comment]
@enj: "just use oc" is what failed for years from a developer experience POV. Feel free to reach to me if you want to pursue. We need login in odo, period.

**Comment by enj** [comment]
> We need login in odo, period.

There is no reason `odo login` cannot call `oc login` under the hood.

If you insist on vendoring an extra 4M LOC, do not expect any support from us.

**Comment by veillard** [comment]
@enj: if you had provided this as an API, which TBH is a fundamental key to enable those 4M LOC as a platform, we would not be there. Give us a supported API to do this, since we don't, well we go with what we can ...

File: cmd/login.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	Token    string
+	CAAuth   string
+	SkipTLS  bool
+)
+
+// versionCmd represents the version command
+var loginCmd = &cobra.Command{
+	Use:   "login",
+	Short: "Login to cluster",
+	Long:  "Login to cluster",
```
Comment (anmolbabu) [inline cmd/login.go]: Will it be better to explain what args are mandatory/mutually exclusive etc... here ?

File: cmd/login.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	Example: `  # Print the client version of Odo
```
Comment (anmolbabu) [inline cmd/login.go]: Is this by mistake?

File: pkg/auth/login.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	"os"
+
+	"github.com/openshift/origin/pkg/oc/cli/login"
+	"k8s.io/kubernetes/pkg/kubectl/genericclioptions"
+
+	"k8s.io/client-go/tools/clientcmd"
+	kclientapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+func Login(server, Username, Password, Token, CAAuth string, SkipTLS bool) error {
```
Comment (anmolbabu) [inline pkg/auth/login.go]: Why params are capitalised?
can it be username, password, token, caAuth instead?

File: cmd/login.go (hunk 1 above)
Comment (surajnarwade) [reply cmd/login.go]: nothing is mandatory, I have updated the example section

**Comment by syamgk** [comment]
there is a error am getting while following the below scenario
1. login to a cluster A
2. logout
3. login to cluster B

```sh
$ oc login https://clusterA:8443 --username=name --password=foo
$ oc logout
$ odo login https://192.168.42.229:8443 --username=name --password=foo
challenger chose not to retry the request
```

**Comment by surajnarwade** [comment]
@syamgk , can you try whole scenario using odo and oc separately ?

Since we are doing any login magic, I think issue is from openshift side ?

**Comment by surajnarwade** [comment]
@anmolbabu added e2e tests

**Comment by syamgk** [comment]
@surajnarwade same scenario works with oc
```
$./oc version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://api.starter-us-west-2.openshift.com:443
openshift v3.11.16
kubernetes v1.11.0+d4cacc0
$./oc login https://devtools-dev.ext.devshift.net:8443 --token=XXXXXXXXXXXXXXXXXXX
Logged into "https://devtools-dev.ext.devshift.net:8443" as "sgk" using the token provided.

You have access to the following projects and can switch between them with 'oc project <projectname>':

  * floc
    just4test

Using project "floc".
$./oc logout
Logged "sgk" out on "https://devtools-dev.ext.devshift.net:8443"
$./oc login https://192.168.42.229:8443  --username=developer --password=foo
Login successful.

You have one project on this server: "myproject"

Using project "myproject".
```

**Comment by kadel** [comment]
It looks like CI checks are in some weird state. @surajnarwade  Can please try to update this branch to retrigger it?

File: .wwhrd.yml [hunk 4]
```diff
@@ -9,9 +9,12 @@ whitelist:
...
   - ISC
+  - MPL-2.0
 
 exceptions:
   # wwhrd currently does not detect ISC which is compatible with Odo so, add it under exceptions to avoid errors due to it being not recognised
   - github.com/davecgh/go-spew/spew/...
   # wwhrd currently does not detect ISC which is compatible with Odo so, add it under exceptions to avoid errors due to it being not recognised
   - github.com/olekukonko/tablewriter
+  - github.com/certifi/gocertifi
+  - github.com/opencontainers/go-digest
```
Comment (kadel) [inline .wwhrd.yml]: Please add comment what licences are those libs and why they are in exceptions.

**Comment by kadel** [comment]
> _No description provided._

no description no link to the issue :-(

**Comment by kadel** [comment]
this is weird, tests are failing on service e2e tests, but all other prs are passing.
This doesn't look like a flaky test.

**Comment by geoand** [comment]
All jobs seem to be passing now

**Comment by surajnarwade** [comment]
@geoand issue that you described is being tracked here https://github.com/redhat-developer/odo/issues/965 and we will take care of it in another PR

**Comment by geoand** [comment]
Thanks for pointing me to the issue Suraj

File: tests/e2e/e2e_test.go [hunk 5]
```diff
@@ -923,4 +923,21 @@ var _ = Describe("updateE2e", func() {
 			EnvVarTest("wildfly-"+appTestName, "binary", "keyvaluekey1value1")
 		})
 	})
+
+	Context("odo login", func() {
+		It("should login with username and password", func() {
+			runCmd("oc logout")
+			runCmd("odo login -u developer -p developer")
```
Comment (mik-dass) [inline tests/e2e/e2e_test.go]: There are no `-u` and `-p` in `odo login`

Comment (surajnarwade) [reply tests/e2e/e2e_test.go]: updated, thanks

Comment (kadel) [reply tests/e2e/e2e_test.go]: Good catch

Comment (mik-dass) [reply tests/e2e/e2e_test.go]: Resolved Great :+1:

File: cmd/login.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	},
+}
+
+func init() {
+	// Add a defined annotation in order to appear in the help menu
+	loginCmd.Annotations = map[string]string{"command": "utility"}
+	loginCmd.SetUsageTemplate(cmdUsageTemplate)
+	loginCmd.Flags().StringVar(&userName, "username", userName, "username, will prompt if not provided")
+	loginCmd.Flags().StringVar(&password, "password", password, "password, will prompt if not provided")
+	loginCmd.Flags().StringVarP(&token, "token", "t", token, "token, will prompt if not provided")
```
Comment (mik-dass) [inline cmd/login.go]: Do we need -t? Because -u and -p are missing

File: cmd/login.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+			util.CheckError(err, "")
+		}
+	},
+}
+
+func init() {
+	// Add a defined annotation in order to appear in the help menu
+	loginCmd.Annotations = map[string]string{"command": "utility"}
+	loginCmd.SetUsageTemplate(cmdUsageTemplate)
+	loginCmd.Flags().StringVar(&userName, "username", userName, "username, will prompt if not provided")
```
Comment (kadel) [inline cmd/login.go]: We are not going to support short flags, like `oc login`?

File: cmd/login.go (hunk 6 above)
Comment (kadel) [reply cmd/login.go]: `oc login` has it exactly the other way. It has  `-u` and `-p` but doesn't have `-t`.
We should follow the same path, so `odo login` is fully compatible with `oc login`.

File: cmd/login.go (hunk 7 above)
Comment (surajnarwade) [reply cmd/login.go]: Added short flags, it was causing problem earlier for me so I removed it for a while

**Comment by surajnarwade** [comment]
@kadel added short flags

**Pull request merged**